src/encode.rs

```diff
@@ -113,373 +87,203 @@ impl<W: io::Write> BitWrite for BitWriter<W> {
+        }
+        Term::Iden => {
+            w.write_bits_be(8, 5)?;
+        }
+        Term::Unit => {
+            w.write_bits_be(9, 5)?;
         }
-        Term::InjL(i) => Ok(writer.write_u8(4, 5)? + encode_natural(i, &mut *writer)?),
-        Term::InjR(i) => Ok(writer.write_u8(5, 5)? + encode_natural(i, &mut *writer)?),
-        Term::Take(i) => Ok(writer.write_u8(6, 5)? + encode_natural(i, &mut *writer)?),
-        Term::Drop(i) => Ok(writer.write_u8(7, 5)? + encode_natural(i, &mut *writer)?),
-        Term::Iden => writer.write_u8(8, 5),
-        Term::Unit => writer.write_u8(9, 5),
         Term::Fail(..) => unimplemented!(),
         Term::Hidden(cmr) => {
-            let mut len = writer.write_u8(6, 4)?;
-            for byte in cmr.as_ref() {
-                len += writer.write_u8(*byte, 8)?;
-            }
-            Ok(len)
+            w.write_bits_be(6, 4)?;
+            encode_hash(cmr.as_ref(), w)?;
         }
-        Term::Witness(..) => writer.write_u8(7, 4),
-        Term::Ext(ref b) => extension::Jet::encode(b, writer),
-        Term::Jet(ref j) => j.encode(writer),
-    }
-}
+        Term::Witness(..) => {
+            w.write_bits_be(7, 4)?;
+        }
+        Term::Ext(ref b) => {
+            Jet::encode(b, w)?;
+        }
+        Term::Jet(ref j) => {
+            j.encode(w)?;
+        }
+    };
 
-pub fn decode_program_no_witness<I: Iterator<Item = u8>, Ext: extension::Jet>(
-    iter: &mut BitIter<I>,
-) -> Result<UntypedProgram<(), Ext>, Error> {
-    let prog_len = decode_natural(&mut *iter, None)?;
+    Ok(())
+}
 
-    // FIXME make this a reasonable limit
-    if prog_len > 1_000_000 {
-        return Err(Error::TooManyNodes(prog_len));
+/// Encode a hash to bits.
+fn encode_hash<W: io::Write>(h: &[u8], w: &mut BitWriter<W>) -> io::Result<()> {
+    for byte in h {
+        w.write_bits_be(*byte as u64, 8)?;
     }
 
-    let mut program = Vec::with_capacity(prog_len);
-    for idx in 0..prog_len {
-        let node = decode_node_no_witness(idx, iter)?;
-        let node = if let Term::Case(i, j) = node {
-            match (&program[idx - i], &program[idx - j]) {
-                (Term::Hidden(..), Term::Hidden(..)) => {
-                    return Err(Error::CaseMultipleHiddenChildren)
-                }
-                (Term::Hidden(..), _) => Term::AssertR(i, j),
-                (_, Term::Hidden(..)) => Term::AssertL(i, j),
-                _ => Term::Case(i, j),
-            }
-        } else {
-            node
-        };
-        program.push(node);
-    }
-
-    Ok(UntypedProgram(program))
+    Ok(())
 }
 
-/// Encode a natural number according to section 7.2.1 of the Simplicity tech
-/// report. Returns the length of the written number, in bits
-pub fn encode_natural<W: BitWrite>(n: usize, writer: &mut W) -> io::Result<usize> {
+/// Encode a natural number to bits.
+pub fn encode_natural<W: io::Write>(n: usize, w: &mut BitWriter<W>) -> io::Result<()> {
```

Comment:
Should all functions return the number of written bits, or is it easier to calculate the total number "higher up" using subtraction?